File: ml-benchmark/synduce/constraints/sortedlist/exists_equal_elements.f
Inductive List = elt Int | cons {Int, List};
TwoList = {List, List};

is_sorted =
  let aux = fix (
    \f: Int -> List -> Bool. \pre: Int. \xs: List.
    match xs with
      elt x -> >= pre x
    | cons {h, t} -> and (>= pre h) (f h t)
    end
  ) in \xs: List.
  match xs with
    elt x -> true
  | cons {h, t} -> aux h t
  end;

is_sorted_pair =
  \p: TwoList. and (is_sorted p.1) (is_sorted p.2);

find = \w: Int. fix (
  \f: List -> Bool. \xs: List.
  match xs with
    elt x -> == x w
  | cons {h, t} -> or (== w h) (f t)
  end
);

spec = \p: TwoList. (fix (
  \f: List -> Bool. \xs: List.
  match xs with
    elt a -> find a p.1
  | cons {h, t} -> or (find h p.1) (f t)
  end
)) p.2;

target = fix (
  \f: TwoList -> Compress TwoList. \p: TwoList.
  match p with
    {elt a, elt b} -> p
  | {elt a, cons {h, t}} ->
    if > a h then p /*Avoid recursions*/ else
      let res = f {elt a, t} in
        {res.1, cons {h, res.2}}
  | {cons {h, t}, elt b} ->
    if > b h then p /*Avoid recursions*/ else
      let res = f {t, elt b} in
        {cons {h, res.1}, res.2}
  | {cons {h1, t1}, cons {h2, t2}} ->
    if > h1 h2 then
      let res = f {t1, p.2} in
        {cons {h1, res.1}, res.2}
    else
      let res = f {p.1, t2} in
        {res.1, cons {h2, res.2}}
  end
);

main = \p: TwoList. if is_sorted_pair p then spec (target p) else false;

File: ml-benchmark/synduce/constraints/sortedlist/exists_equal_elements.ml
type list_ = Elt of int | Cons of int * list_
type twolist = { l1: list_; l2: list_ }

let is_sorted =
  let rec aux pre xs =
    match xs with
    | Elt x -> pre <= x
    | Cons (h, t) -> pre <= h && aux h t
  in
  fun xs ->
    match xs with
    | Elt _ -> true
    | Cons (h, t) -> aux h t

let is_sorted_pair p = is_sorted p.l1 && is_sorted p.l2

let find w =
  let rec f xs =
    match xs with
    | Elt x -> x = w
    | Cons (h, t) -> h = w || f t
  in
  f

let spec p =
  let rec f xs =
    match xs with
    | Elt a -> find a p.l1
    | Cons (h, t) -> find h p.l1 || f t
  in
  f p.l2

let rec target p =
  match p with
  | { l1; l2 } ->
      (match l1 with
       | Elt a ->
           (match l2 with
            | Elt _ -> p
            | Cons (h, t) ->
                if a > h then p
                else
                  let res = target { l1 = Elt a; l2 = t } in
                  { l1 = res.l1; l2 = Cons (h, res.l2) })
       | Cons (h, t) ->
           (match l2 with
            | Elt b ->
                if b > h then p
                else
                  let res = target { l1 = t; l2 = Elt b } in
                  { l1 = Cons (h, res.l1); l2 = res.l2 }
            | Cons (h2, t2) ->
                if h > h2 then
                  let res = target { l1 = t; l2 = l2 } in
                  { l1 = Cons (h, res.l1); l2 = res.l2 }
                else
                  let res = target { l1 = l1; l2 = t2 } in
                  { l1 = res.l1; l2 = Cons (h2, res.l2) }))

let program p =
  if is_sorted_pair p then spec (target p) else false